src/segment.rs

```diff
@@ -63,44 +62,597 @@ impl<'a> Segment<'a> {
+        }
+        Ok(nested_pointer)
+    }
+
+    pub fn datastore_nested(
+        &self,
+        transaction: impl AsRef<Transaction>,
+        params: impl AsRef<DatastoreParams>,
+    ) -> Result<Self> {
+        let inner = self.inner.ok_or_else(|| {
+            error!("Could not create external segment due to invalid parent segment");
+            Error::SegmentStartError
+        })?;
+        let transaction = transaction.as_ref();
+        let params = params.as_ref();
+        let nested_pointer = Self::datastore(transaction, params)?;
+        // If result is ok, then guaranteed there is some pointer
+        let nested_inner = nested_pointer.inner.unwrap();
+        unsafe {
+            ffi::newrelic_set_segment_parent(nested_inner, inner);
+        }
+        Ok(nested_pointer)
+    }
+
+    pub fn external_nested(
+        &self,
+        transaction: impl AsRef<Transaction>,
+        params: impl AsRef<ExternalParams>,
+    ) -> Result<Self> {
+        let inner = self.inner.ok_or_else(|| {
+            error!("Could not create external segment due to invalid parent segment");
+            Error::SegmentStartError
+        })?;
+        let transaction = transaction.as_ref();
+        let params = params.as_ref();
+        let nested_pointer = Self::external(transaction, params)?;
+        // If result is ok, then guaranteed there is some pointer
+        let nested_inner = nested_pointer.inner.unwrap();
+        unsafe {
+            ffi::newrelic_set_segment_parent(nested_inner, inner);
+        }
+        Ok(nested_pointer)
+    }
+
+    #[cfg(feature = "distributed_tracing")]
+    pub fn distributed_trace(
+        &self,
+        transaction: impl AsRef<Transaction>,
+    ) -> Option<String> {
+        let transaction = transaction.as_ref();
+        self.inner.map(|pointer| {
+            let payload = FreeableString::new(unsafe {
+                ffi::newrelic_create_distributed_trace_payload_httpsafe(transaction.inner, pointer)
+            });
+            payload.convert()
+        })
+    }
+
+    pub fn end(
+        &mut self,
+        transaction: impl AsRef<Transaction>,
+    ) {
+        if let Some(mut inner) = self.inner {
+            let transaction = transaction.as_ref();
+            unsafe {
+                ffi::newrelic_end_segment(transaction.inner, &mut inner);
+            }
+            debug!("Ended segment");
+            self.inner = None;
+        }
+    }
+}
+
+unsafe impl Send for SegmentPointer {}
+
+unsafe impl Sync for SegmentPointer {}
+
+
+/// A segment with a reference to a transaction via the `AsRef` trait.
+///
+/// Use segments to instrument transactions with greater granularity.
+///
+/// `ReferencingSegment` can be nested by calling the various `_nested` methods on
+/// an existing borrowing segment.
+pub struct ReferencingSegment<T: AsRef<Transaction> + Clone> {
+    /// A reference to a parental transaction as mediated by the `Borrow` trait.
```

Comment:
```suggestion
    /// A reference to a parental transaction as mediated by the `AsRef` trait.
```

Reply:
We should be shortly provided no impediments in other work get in the way. I will report back when I have confirmation that things are running smoothly.

Reply:
Sorry it's been so long. I had been moved onto other things. I can now confirm that we have been running this in production without issue for some time in an Actixweb api project.

Reply:
Great, thanks! I'll merge now.